src/System.Text.Json/src/JsonReflectionEmitMaterializer.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Reflection;
+using System.Reflection.Emit;
+
+namespace System.Text.Json.Serialization
+{
+    internal class JsonReflectionEmitMaterializer : JsonMemberBasedClassMaterializer
+    {
+        public override JsonClassInfo.ConstructorDelegate CreateConstructor(Type type)
+        {
+            ConstructorInfo realMethod = type.GetConstructor(BindingFlags.Public | BindingFlags.NonPublic| BindingFlags.Instance, binder: null, Type.EmptyTypes, modifiers: null);
+            if (realMethod == null)
+                return null;
+
+            var dynamicMethod = new DynamicMethod(
+                realMethod.Name,
+                type,
+                Type.EmptyTypes,
+                typeof(JsonReflectionEmitMaterializer).Module,
+                skipVisibility: true);
+
+            if (dynamicMethod != null)
+            {
+                ILGenerator generator = dynamicMethod?.GetILGenerator();
+                if (generator != null)
+                {
+
+                    generator.Emit(OpCodes.Newobj, realMethod);
+                    generator.Emit(OpCodes.Ret);
+
+                    var result = (JsonClassInfo.ConstructorDelegate)dynamicMethod.CreateDelegate(typeof(JsonClassInfo.ConstructorDelegate));
+                    return result;
+                }
+            }
+
+            throw new InvalidOperationException(SR.Format(SR.SerializationUnableToCreateDynamicMethod, $"{type.FullName}.{realMethod.Name}"));
+        }
+
+        public override JsonPropertyInfo<TValue>.GetterDelegate CreateGetter<TValue>(PropertyInfo propertyInfo)
+        {
+            MethodInfo realMethod = propertyInfo.GetGetMethod();
+            Debug.Assert(realMethod != null);
+
+            var dynamicMethod = new DynamicMethod(
+                realMethod.Name,
+                typeof(TValue),
+                new Type[] { typeof(object) },
+                typeof(JsonReflectionEmitMaterializer).Module,
+                skipVisibility: true);
+
+            if (dynamicMethod != null)
+            {
+
+                ILGenerator generator = dynamicMethod?.GetILGenerator();
+                if (generator != null)
+                {
+                    generator.Emit(OpCodes.Ldarg_0);
+                    generator.EmitCall(OpCodes.Callvirt, realMethod, null);
+                    generator.Emit(OpCodes.Ret);
+
+                    var result = (JsonPropertyInfo<TValue>.GetterDelegate)dynamicMethod.CreateDelegate(typeof(JsonPropertyInfo<TValue>.GetterDelegate));
+                    return result;
+                }
```

Comment:
Why do we need reflection emit for this?  Can we not use MethodInfo.CreateDelegate? e.g.
```C#
return (JsonPropertyInfo<TValue>.GetterDelegate)realMethod.CreateDelegate(typeof(JsonPropertyInfo<TValue>.GetterDelegate));
```

Reply:
Reflection emit is used for performance. If not available then we use standard reflection which uses the getter\setter's MethodInfo.

Reply:
Are you sure it's faster when compared with MethodInfo.CreateDelegate?

Reply:
I'll have a new commit for this shortly.

I thought you meant just MethodInfo.Invoke. Yes if we get a real delegate to the getter\setter then it should be just as fast. The tricky part is that I need both the poco type and the property type to create that delegate - I have this working now by changing the JsonPropertyInfo<TProperty> to JsonPropertyInfo<TClass, TProperty> and building the delegate off of TClass and TProperty. 

However I had an optimization to prevent instances of our internal converters (e.g. `JsonPropertyInfoInt16`) from being created by deriving from JsonPropertyInfo<TProperty>. That will no longer work since I don't want to create converter instances that depend on the TClass. However, there may be a better way of handling converters using delegates to each method (read, write) instead that would avoid creating a bunch of unnecessary objects - I'll add a todo for that.